// @replit/crosis/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloseCode = exports.FetchConnectionMetadataError = exports.ConnectionState = void 0;
var ConnectionState;
(function (ConnectionState) {
    ConnectionState[ConnectionState["CONNECTING"] = 0] = "CONNECTING";
    ConnectionState[ConnectionState["CONNECTED"] = 1] = "CONNECTED";
    ConnectionState[ConnectionState["DISCONNECTED"] = 2] = "DISCONNECTED";
})(ConnectionState = exports.ConnectionState || (exports.ConnectionState = {}));
var FetchConnectionMetadataError;
(function (FetchConnectionMetadataError) {
    FetchConnectionMetadataError["Aborted"] = "Aborted";
    FetchConnectionMetadataError["Retriable"] = "Retriable";
})(FetchConnectionMetadataError = exports.FetchConnectionMetadataError || (exports.FetchConnectionMetadataError = {}));
exports.CloseCode = {
    INVALID_UPSTREAM_RESPONSE: 1014,
    POLICY_VIOLATION: 1008,
    FIREWALL_DENIED: 4000,
    TRY_ANOTHER_MACHINE: 4001,
    USER_ERROR: 4002,
};
